core/node/rpc/sync/operation.go

```diff
@@ -78,15 +78,66 @@ func NewStreamsSyncOperation(
 	}, nil
 }
 
+// sendMsgClientLoop sends msg to res and cancels the sync operation and returns in case of an error.
+// res is closed when the sync operation is cancelled.
+func (syncOp *StreamSyncOperation) sendMsgClientLoop(
+	res *connect.ServerStream[SyncStreamsResponse],
+	sendQueue <-chan *syncStreamsResponseReply,
+) {
+	log := dlog.FromCtx(syncOp.ctx).With("syncId", syncOp.SyncID)
+
+	for msg := range sendQueue {
+		if err := res.Send(msg.msg); err != nil {
+			syncOp.cancel()
+			log.Debug("Unable to send msg to client", "err", err)
+			msg.err <- err
+			close(msg.err)
+			return
+		}
+		close(msg.err)
+	}
+}
+
+type syncStreamsResponseReply struct {
+	msg *SyncStreamsResponse
+	err chan error
+}
+
+// forwardMessageToClient writes msg to sendQueue which is picked up by sendMessagesToClient. sendMessagesToClient
+// writes the result to msg.err. If no results is read within the given timeout forwardMessageToClient returns an
+// Err_DEADLINE_EXCEEDED error. The caller is expected to cancel the sync operation.
+func (syncOp *StreamSyncOperation) forwardMessageToClient(
+	timeout time.Duration,
+	sendQueue chan<- *syncStreamsResponseReply,
+	msg *syncStreamsResponseReply,
+) error {
+	sendTimeout := time.NewTimer(timeout)
+
+	select {
+	case sendQueue <- msg:
+		select {
+		case err := <-msg.err:
+			if !sendTimeout.Stop() {
+				<-sendTimeout.C
+			}
+			return err
+		case <-sendTimeout.C:
+			return RiverError(Err_DEADLINE_EXCEEDED, "timeout sending sync stream update to client")
```

Comment:
Is there any information difference between this and the error on line 128 that would cause it to make sense to distinguish tme?

Reply:
I don't think it's relevant, in both case the send operation is taking too long due to a problem deep in the stack. In the first case the work loop accepted the request (should always be the case because otherwise the previous msg was stuck and should have timed out) and in the second case the send operation took too long.